[DotNetCore] Display NuGet package dependency diagnostics

mrward opened the pull request:
NuGet package diagnostic warnings are now shown in the Solution window.
The diagnostic warnings are shown underneath the NuGet package with
a warning icon. Hovering over the warning icon will show a tooltip.
If the NuGet package has only one diagnostic child then the tooltip message will show the
diagnostic message, otherwise the tooltip will indicate that it should be expanded to see the diagnostic. Diagnostics can occur at any level in the dependencies tree so packages that have a diagnostic somewhere below in their hierarchy will show a warning icon.

Fixes VSTS #549720 - Display NuGet package dependency diagnostics in
Solution window

Needs a design review/approval before being merged - there is a separate design VSTS item opened for this. At minimum probably needs a new icon for the diagnostic tree node.

<img width="371" alt="nugetwarnings1" src="https://user-images.githubusercontent.com/372361/40647255-537c5d8a-6323-11e8-838d-a8049d0cf003.png">

<img width="382" alt="nugetwarnings2" src="https://user-images.githubusercontent.com/372361/40647265-57e4e3ec-6323-11e8-8074-78da78e52da7.png">

<img width="513" alt="nugetwarningnested3" src="https://user-images.githubusercontent.com/372361/40647282-5daf3e80-6323-11e8-9b19-e4d4fca4b495.png">

mrward committed:
[DotNetCore] Add tests for dependency node generation

Added tests for the nodes generated for the Dependencies folder. Tests
the for the nodes added for the SDK and NuGet folders have been added.

mrward committed:
[DotNetCore] Display NuGet package dependency diagnostics

NuGet package diagnostic warnings are now shown in the Solution window.
The diagnostic warnings are shown underneath the NuGet package with
a warning icon. Hovering over the warning icon will show a tooltip.
If the NuGet package is at the top level then a warning icon is
shown for the NuGet package and the tooltip message will show the
diagnostic message if only one diagnostic is associated with the
NuGet package.

Note that diagnostic messages can be associated with a NuGet package
at any level in the hierarchy. Whilst these are displayed in the
Solution window it is difficult to find them since there is no
indicating there is a warning higher up the tree.

Fixes VSTS #549720 - Display NuGet package dependency diagnostics in
Solution window

mrward committed:
[DotNetCore] Show warning icon for indirect NuGet diagnostics

Warning icons are now shown in the Solution pad for all higher
NuGet packages in the Dependencies tree structure. This makes it
easier to find the warning in the tree without having to expand all
items. Now you can follow the warning icons down the tree until you
find the NuGet diagnostic for a particular NuGet package.

mrward committed:
[DotNetCore] Build package dependency hierarchy on non-UI thread

The dependency tree information was being built on the UI thread
after the package dependencies were returned back from MSBuild, which
was run on a background thread.

slluis commented:
Can you provide a project that can be used to test the PR and review the UI?

mrward commented:
Added a url to a [github repository](https://github.com/mrward/TestNetStandardNuGetWarnings) with projects that have diagnostic warnings. Includes the ones that were used in the unit tests.

vancura commented:
> There is a separate design VSTS item opened for this.

Where, please?

mrward commented:
Design issue: https://devdiv-design.visualstudio.com/DevDiv%20Design/_workitems/edit/2937

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.NodeBuilders/PackageDependencyInfo.cs:
```diff
@@ -0,0 +1,72 @@
+	class PackageDependencyInfo
+	{
+		ImmutableArray<PackageDependencyInfo> dependencies = ImmutableArray<PackageDependencyInfo>.Empty;
+		PackageDependency dependency;
```
```suggestion
		readonly PackageDependency dependency;
```

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.NodeBuilders/PackageDependencyInfo.cs:
```diff
@@ -0,0 +1,72 @@
+			get { return dependencies; }
+		}
+
+		public void AddChild (PackageDependencyInfo dependency)
```
Style nit: this clashes with the field name, maybe it would be better to have it named differently for clarity.

vancura commented:
Icons: https://github.com/xamarin/icons/commit/c46eb82866eb9f99aa9221467f2a17618e80aca5

mrward committed:
[DotNetCore] Address pull request feedback

Make field readonly. Change method argument to avoid clash with
field in class.

mrward committed:
[NuGet] Add NuGet package warning icon in solution window

NuGet restore warning codes (e.g. NU1603) now have a yellow coloured
icon compared with the other package dependency icons in the
Solution window.

mrward commented:
Addressed PR feedback and added new icons. Some new screenshots:

<img width="463" alt="darktheme" src="https://user-images.githubusercontent.com/372361/47225482-18e57680-d3b6-11e8-81f4-d8931c744a7a.png">
<img width="464" alt="lighttheme" src="https://user-images.githubusercontent.com/372361/47225486-1c78fd80-d3b6-11e8-83c7-916984dd6606.png">
<img width="439" alt="lighttheme2" src="https://user-images.githubusercontent.com/372361/47225494-1edb5780-d3b6-11e8-985f-b7c6684339ef.png">
